src/net/posix-stack.cc

```diff
@@ -769,21 +775,29 @@ class posix_udp_channel : public udp_channel_impl {
             resolve_outgoing_address(_dst);
         }
     };
+    static bool eligible_to_ip_pktinfo(const socket_address& addr) {
+        return addr.family() == AF_INET || addr.family() == AF_INET6;
+    }
     pollable_fd _fd;
     socket_address _address;
     recv_ctx _recv;
     send_ctx _send;
     bool _closed;
 public:
     posix_udp_channel(const socket_address& bind_address)
-            : _closed(false) {
+            : _recv(eligible_to_ip_pktinfo(bind_address) || bind_address.is_unspecified()),
+            _closed(false) {
         auto sa = bind_address.is_unspecified() ? socket_address(inet_address(inet_address::family::INET)) : bind_address;
         file_desc fd = file_desc::socket(sa.u.sa.sa_family, SOCK_DGRAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
-        fd.setsockopt(SOL_IP, IP_PKTINFO, true);
-        if (engine().posix_reuseport_available()) {
-            fd.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1);
+
+        if (eligible_to_ip_pktinfo(sa)) {
+            fd.setsockopt(SOL_IP, IP_PKTINFO, true);
+            if (engine().posix_reuseport_available()) {
```

Comment:
Why did you move this case inside the if(eligible)? is it because SO_REUSEPORT happens to work on the same type of sockets that support IP_PKTINFO? But the SOL_SOCKET instead of SOL_IP suggests it should have worked on Unix domain sockets as well (although perhaps not be especially useful?)

Reply:
> is it because SO_REUSEPORT happens to work on the same type of sockets that support IP_PKTINFO

yes, i should have named `eligible_to_ip_pktinfo` somehow better